pkg/k8s/watchers/pod.go

```diff
@@ -164,164 +150,58 @@ func hostPortServiceName(pod *slim_corev1.Pod, port uint16, cluster string) *loa
-		podStoreOnce sync.Once
-	)
-	watchNodePods := func() context.CancelFunc {
-		ctx, cancel := context.WithCancel(context.Background())
-		var synced atomic.Bool
+func (k *K8sPodWatcher) podsInit(ctx context.Context) {
+	var synced atomic.Bool
 
-		go func() {
-			_, initWatch := k.pods.Initialized(k.db.ReadTxn())
-			select {
-			case <-ctx.Done():
-			case <-initWatch:
-			}
-			synced.Store(true)
-		}()
-
-		go func() {
-			pods := make(map[types.NamespacedName]*slim_corev1.Pod)
-			wtxn := k.db.WriteTxn(k.pods)
-			changeIter, err := k.pods.Changes(wtxn)
-			wtxn.Commit()
-			if err != nil {
-				return
-			}
+	go func() {
+		_, initWatch := k.pods.Initialized(k.db.ReadTxn())
+		select {
+		case <-ctx.Done():
+		case <-initWatch:
+		}
+		synced.Store(true)
+	}()
 
-			for {
-				changes, watch := changeIter.Next(k.db.ReadTxn())
-				for change := range changes {
-					pod := change.Object.Pod
-					name := types.NamespacedName{Namespace: pod.Namespace, Name: pod.Name}
-					if !change.Deleted {
-						oldPod := pods[name]
-						if oldPod == nil {
-							k.addK8sPodV1(pod)
-						} else {
-							k.updateK8sPodV1(oldPod, pod)
-						}
-						k.k8sResourceSynced.SetEventTimestamp(podApiGroup)
-						pods[name] = pod
+	go func() {
+		pods := make(map[types.NamespacedName]*slim_corev1.Pod)
+		wtxn := k.db.WriteTxn(k.pods)
+		changeIter, err := k.pods.Changes(wtxn)
+		wtxn.Commit()
+		if err != nil {
+			return
+		}
 
+		for {
+			changes, watch := changeIter.Next(k.db.ReadTxn())
+			for change := range changes {
+				pod := change.Object.Pod
+				name := types.NamespacedName{Namespace: pod.Namespace, Name: pod.Name}
+				if !change.Deleted {
+					oldPod := pods[name]
+					if oldPod == nil {
+						k.addK8sPodV1(pod)
 					} else {
-						k.deleteK8sPodV1(pod)
-						k.k8sResourceSynced.SetEventTimestamp(podApiGroup)
-						delete(pods, name)
+						k.updateK8sPodV1(oldPod, pod)
 					}
-				}
-				select {
-				case <-ctx.Done():
-					return
-				case <-watch:
+					k.k8sResourceSynced.SetEventTimestamp(podApiGroup)
+					pods[name] = pod
+
+				} else {
+					k.deleteK8sPodV1(pod)
+					k.k8sResourceSynced.SetEventTimestamp(podApiGroup)
+					delete(pods, name)
 				}
 			}
-		}()
-
-		k.k8sResourceSynced.BlockWaitGroupToSyncResources(ctx.Done(), nil, synced.Load, resources.K8sAPIGroupPodV1Core)
-		apiOnce.Do(func() {
-			asyncControllers.Done()
-			k.k8sAPIGroups.AddAPI(resources.K8sAPIGroupPodV1Core)
-		})
-		return cancel
-	}
-
-	// We will watch for pods on the entire cluster to keep existing
-	// functionality untouched. If we are running with CiliumEndpoint CRD
-	// enabled then it means that we can simply watch for pods that are created
-	// for this node. Similarly, we don't need to watch for all pods when the
-	// running in KVStore mode, as in that case we just rely on the KVStore data
-	// from the beginning.
-	if !option.Config.DisableCiliumEndpointCRD || option.Config.KVstoreEnabled() {
-		watchNodePods()
-		return
-	}
```

Comment:
There seem to be currently two possibilities in which the CiliumEndpoint CRD may be disabled by the user, hence falling back here to watching all pods rather than the local ones only:

1. Cilium is running in kvstore mode; in this case, the fallback logic is not necessary, and we can drop it (it is also already currently disabled), because we rely on the information retrieved from etcd.
2. NetworkPolicy support is disabled (related: 19fe642a4fa8ed8c8a5980c97cbd633f7a1ea2c1). @dlapcevic Could you please confirm whether it is necessary to watch all pods in this case? If so, I'll have to rework this commit to preserve the current logic for that use-case. FWIW, I'm not seeing any CI failure that may indicate problems, but I'm unsure about the coverage in that area.